### ISSUE TITLE
Support for AbortSignal?

### ISSUE THREAD

**Issue by spoike** [petkaantonov/bluebird#1528]
Would it make sense to include support for `AbortController`/`AbortSignal` in the API or at least document the uses of it?

## What is `AbortSignal`?

There seems to be a growing consensus about the new DOM api for aborting fetch requests with the `AbortSignal` interface. Refs:

* https://developers.google.com/web/updates/2017/09/abortable-fetch
* https://developer.mozilla.org/en-US/docs/Web/API/AbortController
* https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal

The interface for `AbortSignal` is rather simple, on top of extending `EventTarget`:

* `aborted` boolean read-only that tells if process has been aborted or not
* `onabort` event handler when abort has been issued (from e.g. AbortController)

The apparent benefit from this pattern is that you can cancel multiple processes at once.

## What can be done in Bluebird?

* Extend functions that do serial iterations so they can be cancelled through an `AbortSignal` by providing the signal as argument. E.g. `Promise.each([p1, p2, p3], callback, signal)`, `reduce`, etc.
* Documenting the use of the new DOM api when creating promises:
```
const controller = new AbortController();
const signal = controller.signal;

new Promise((resolve, reject) => {
  const to = setTimeout(() => { resolve('boop'); }, 300);

  signal.addEventListener('abort', () => {
    clearTimeout(to);
    reject('cancelled');
  });

  console.log('beep');
});

controller.abort();
// OUTPUT:
// beep
// cancelled
```
* Provide a simple AbortSignal implementation but with an `abort` method included.

**Comment by benjamingr** [comment]
> Would it make sense to include support for AbortController/AbortSignal in the API or at least document the uses of it?

Well, thanks for opening the issue but I don't think so. AbortController is a compromise reached because the lack of a better API - Bluebird ships with promise cancellation.

Note that you can use AbortController with bluebird just fine and mix and match since an AbortController is external to promises - but it can't do things like abort coroutines in the middle or promise chains and it doesn't have resource cleanup or disinterest semantics.

Basically, you can use promise cancellation and trigger `.cancel` with `AbortController.signal`

**Comment by benjamingr** [comment]
By the way, we're (Node.js) discussing this tomorrow at the Node.js collaborator summit (use cases in https://github.com/nodejs/promise-use-cases/tree/master/use-cases and some of them cover cancellation).

Feel free to check out the repo and weigh in.

**Comment by MattiasBuelens** [comment]
FWIW: I've written some adapter functions to convert between a cancellable promise and a promise with an `AbortController` over at [bluebird-abort-signal-adapter](https://github.com/MattiasBuelens/bluebird-abort-signal-adapter). No usage instructions yet, but [the tests](https://github.com/MattiasBuelens/bluebird-abort-signal-adapter/tree/7f29112ae10e24a25a8d961377cb8647459b9a46/test) should give you an idea of how to use it.

If you're interested, I can find some time to put this on npm.
